pragma-node/src/infra/repositories/onchain_repository/history.rs

```diff
@@ -70,3 +103,175 @@ pub async fn get_historical_aggregated_entries(
+            normalize_entries_to_decimals(base_entries, base_decimals, quote_decimals);
+        (
+            convert_entries_via_quote(normalized_base_entries, quote_entries, quote_decimals)?,
+            quote_decimals,
+        )
+    } else {
+        let normalized_quote_entries =
+            normalize_entries_to_decimals(quote_entries, quote_decimals, base_decimals);
+        (
+            convert_entries_via_quote(base_entries, normalized_quote_entries, base_decimals)?,
+            base_decimals,
+        )
+    };
+
+    Ok((rebased_entries, decimals))
+}
+
+/// Wrapper around the [normalize_to_decimals] function. It will be applied to all
+/// entries median_price.
+fn normalize_entries_to_decimals(
+    entries: Vec<HistoricalEntryRaw>,
+    base_decimals: u32,
+    quote_decimals: u32,
+) -> Vec<HistoricalEntryRaw> {
+    entries
+        .into_iter()
+        .map(|mut entry| {
+            entry.median_price =
+                normalize_to_decimals(entry.median_price, base_decimals, quote_decimals);
+            entry
+        })
+        .collect()
+}
+
+/// Wrapper around the [convert_via_quote] function. It will be applied to all
+/// normalized_entries median_price.
+fn convert_entries_via_quote(
+    normalized_entries: Vec<HistoricalEntryRaw>,
+    quote_entries: Vec<HistoricalEntryRaw>,
+    decimals: u32,
+) -> Result<Vec<HistoricalEntryRaw>, InfraError> {
+    normalized_entries
+        .into_iter()
+        .zip(quote_entries)
+        .map(|(base_entry, quote_entry)| {
+            let converted_price = convert_via_quote(
+                base_entry.median_price.clone(),
+                quote_entry.median_price.clone(),
+                decimals,
+            )?;
+            combine_entries(&base_entry, &quote_entry, converted_price)
+        })
+        .collect()
+}
+
+/// Given two entries, determine what should be the resulted pair_id, timestamp
+/// & sources. Returns the new entry afterwards.
+fn combine_entries(
+    base_entry: &HistoricalEntryRaw,
+    quote_entry: &HistoricalEntryRaw,
+    converted_price: BigDecimal,
+) -> Result<HistoricalEntryRaw, InfraError> {
+    let min_timestamp = std::cmp::max(
+        base_entry.timestamp.and_utc().timestamp(),
+        quote_entry.timestamp.and_utc().timestamp(),
+    );
+    let num_sources = std::cmp::max(
+        base_entry.nb_sources_aggregated,
+        quote_entry.nb_sources_aggregated,
+    );
+    let new_timestamp = DateTime::from_timestamp(min_timestamp, 0)
+        .ok_or(InfraError::InvalidTimestamp(format!(
+            "Cannot convert to DateTime: {min_timestamp}"
+        )))?
+        .naive_utc();
+
+    Ok(HistoricalEntryRaw {
+        pair_id: construct_new_pair_id(&base_entry.pair_id, &quote_entry.pair_id)?,
+        timestamp: new_timestamp,
+        median_price: converted_price,
+        nb_sources_aggregated: num_sources,
+    })
+}
+
+fn construct_new_pair_id(base_pair_id: &str, quote_pair_id: &str) -> Result<String, InfraError> {
```

Comment:
move that to utils and see my first comment, this should live in a `Pair` impl

Reply:
removed and used `currency_pair_to_pair_id`